cdap-credential-ext-gcp-wi/src/main/java/io/cdap/cdap/security/spi/credential/GcpWorkloadIdentityCredentialProvider.java

```diff
@@ -0,0 +1,326 @@
+    }
+  }
+
+  private String getK8sServiceAccountToken(String namespace,
+      String serviceAccountName, String audience) throws ApiException, IOException {
+
+    // Create the TokenRequestSpec with the specified audience.
+    V1TokenRequestSpec v1TokenRequestSpec = new V1TokenRequestSpec()
+        .audiences(Collections.singletonList(audience))
+        .expirationSeconds(3600L);
+
+    AuthenticationV1TokenRequest authenticationV1TokenRequest = new AuthenticationV1TokenRequest()
+        .apiVersion("authentication.k8s.io/v1").kind("TokenRequest").spec(v1TokenRequestSpec);
+
+    V1ObjectMeta serviceAccountMetadata = new V1ObjectMeta()
+        .name(serviceAccountName).namespace(namespace);
+    authenticationV1TokenRequest.setMetadata(serviceAccountMetadata);
+
+    CoreV1Api coreV1Api = new CoreV1Api(getApiClient());
+    authenticationV1TokenRequest = coreV1Api.createNamespacedServiceAccountToken(serviceAccountName,
+        namespace, authenticationV1TokenRequest, null, null,
+        null, null);
+
+    return authenticationV1TokenRequest.getStatus().getToken();
+  }
+
+  private String exchangeTokenViaSts(String token, String scopes, String audience)
+      throws IOException {
+
+    SecurityTokenServiceRequest securityTokenServiceRequest =
+        new SecurityTokenServiceRequest(GrantType.TOKEN_EXCHANCE, audience, scopes,
+            TokenType.ACCESS_TOKEN, TokenType.JWT, token);
+    String securityTokenServiceRequestJson = GSON.toJson(securityTokenServiceRequest);
+    URL url = new URL(SecurityTokenServiceRequest.STS_ENDPOINT);
+
+    Map<String, String> headers = new HashMap<>();
+    headers.put(HttpHeaders.CONTENT_TYPE, "application/json");
+    return executeHttpPostRequest(url, securityTokenServiceRequestJson, headers);
+  }
+
+  /**
+   * Executes a http post request with the specified parameters.
+   */
+  @VisibleForTesting
+  String executeHttpPostRequest(URL url, String body, Map<String, String> headers)
+      throws IOException {
+
+    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+    connection.setRequestMethod(HttpMethod.POST);
+    connection.setUseCaches(false);
+    for (Map.Entry<String, String> entry : headers.entrySet()) {
+      connection.setRequestProperty(entry.getKey(), entry.getValue());
+    }
+    connection.setDoOutput(true);
+
+    // Write the request body to the output stream
+    try (DataOutputStream outputStream = new DataOutputStream(connection.getOutputStream())) {
+      outputStream.writeBytes(body);
+      outputStream.flush();
+    }
+
+    StringBuilder response = new StringBuilder();
+    try (BufferedReader in = new BufferedReader(
+        new InputStreamReader(connection.getInputStream()))) {
+      String inputLine;
+      while ((inputLine = in.readLine()) != null) {
+        response.append(inputLine);
+      }
+    }
+    return response.toString();
+  }
+
+  private String getIamServiceAccountToken(String token, String scopes,
```

Comment:
nit: `fetchIamServiceAccountToken` instead, since `get` generally implies a low-cost call, and `fetch` implies a higher-cost call (e.g. network call).

Reply:
done.